experimental/hip/hip_driver.c

```diff
@@ -0,0 +1,438 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <stdint.h>
+#include <string.h>
+
+#include "experimental/hip/api.h"
+#include "experimental/hip/dynamic_symbols.h"
+#include "experimental/hip/status_util.h"
+#include "iree/base/api.h"
+#include "iree/base/assert.h"
+#include "iree/base/tracing.h"
+#include "iree/hal/api.h"
+
+// Maximum device name length supported by the HIP HAL driver.
+#define IREE_HAL_HIP_MAX_DEVICE_NAME_LENGTH 128
+
+// Utility macros to convert between HIPDevice and iree_hal_device_id_t.
+#define IREE_HIPDEVICE_TO_DEVICE_ID(device) (iree_hal_device_id_t)((device) + 1)
+#define IREE_DEVICE_ID_TO_HIPDEVICE(device_id) (hipDevice_t)((device_id)-1)
+
+typedef struct iree_hal_hip_driver_t {
+  // Abstract resource used for injecting reference counting and vtable;
+  // must be at offset 0.
+  iree_hal_resource_t resource;
+
+  iree_allocator_t host_allocator;
+
+  // Identifier used for registering the driver in the IREE driver registry.
+  iree_string_view_t identifier;
+  // HIP driver API dynamic symbols to interact with the HIP system.
+  iree_hal_hip_dynamic_symbols_t HIP_symbols;
```

Comment:
Since ICCL is our goal I didn't include RCCL for HIP backend now because we'd have to work to remove it. This can help drive ICCL decisions as well.

Reply:
I would check with @antiagainst since ICCL might be a different level of abstraction so both rccl and iccl could co-exist and give us the most flexibility.